File: app/src/main/java/com/stho/software/nyota/universe/AbstractUniverseInitializer.java
package com.stho.software.nyota.universe;

import com.stho.software.nyota.utilities.Degree;
import com.stho.software.nyota.utilities.Hour;

abstract class AbstractUniverseInitializer {
    /**
     * Singleton
     */
    protected Universe universe;

    AbstractUniverseInitializer(Universe universe) {
        this.universe = universe;
    }

    Star newStar(String name, String symbol, String ascension, String declination, double brightness) {
        return newStar(name, symbol, Hour.fromHour(ascension), Degree.fromDegree(declination), brightness);
    }

    private Star newStar(String name, String symbol, Hour ascension, Degree declination, double brightness) {
        if (universe.stars.containsKey(name)) {
            return universe.stars.get(name);
        }
        else {
            Star star = new Star(name, symbol, ascension.toDegree(), declination.toDegree(), brightness);
            universe.stars.put(star.getName(), star);
            return star;
        }
    }

    Constellation newConstellation(String name, int imageId) {
        Constellation constellation = new Constellation(name, imageId);
        universe.constellations.put(name, constellation);
        return constellation;
    }

    @SuppressWarnings({"UnusedReturnValue", "SameParameterValue"})
    Satellite newSatellite(String name, String elements) {
        Satellite satellite = new Satellite(name, elements);
        universe.satellites.put(name, satellite);
        return satellite;
    }

    @SuppressWarnings({"UnusedReturnValue", "SameParameterValue"})
    Special newSpecialElement(String name, Hour ascension, Degree declination) {
        Special special = new Special(name, ascension.toDegree(), declination.toDegree());
        universe.specials.add(special);
        return special;
    }

    void asVIP(Star star) {
        if (!universe.vip.contains(star))
            universe.vip.add(star);
    }
}

File: app/src/main/java/com/stho/nyota/sky/universe/AbstractUniverseInitializer.kt
package com.stho.nyota.sky.universe

import com.stho.nyota.sky.utilities.Degree
import com.stho.nyota.sky.utilities.Hour
import java.lang.Exception


abstract class AbstractUniverseInitializer(protected var universe: Universe) {

    fun getStar(hd: Int): Star =
        universe.stars[hd] ?: throw Exception("Star $hd was not initialized yet")

    fun getStar(starName: String): Star =
        universe.stars[starName] ?: throw Exception("Star $starName was not initialized yet")

    fun getStar(rank: Int, symbol: Symbol): Star =
        universe.constellations[rank][symbol] // throws an exception if rank or symbol were not initialized

    fun getConstellation(rank: Int): Constellation =
        universe.constellations[rank] // throws an exception if rank was not initialized

    fun newGalaxy(name: String, imageId: Int, ascension: String, declination: String, brightness: Double, distance: Double): Galaxy =
        Galaxy(name, imageId, Hour.fromHour(ascension).angleInDegree, Degree.fromDegree(declination).angleInDegree, brightness, distance).also {
            universe.galaxies.add(it)
        }

    fun newAnything(name: String, ascension: String, declination: String, brightness: Double, distance: Double): Anything =
        Anything(name, Hour.fromHour(ascension).angleInDegree, Degree.fromDegree(declination).angleInDegree, brightness, distance).also {
            universe.any.add(it)
        }

    fun newHint(description: String, at: IElement): Hint =
        Hint.create(description, at).also {
            universe.hints.add(it)
        }

    fun newHint(description: String, from: Star, to: Star): Hint =
        Hint.create(description, from, to).also {
            universe.hints.add(it)
        }

    fun newTriangle(description: String, one: Star, two: Star, three: Star): Hint =
        Hint.create(description, one, two, three).also {
            universe.hints.add(it)
        }

    fun newRectangle(description: String, one: Star, two: Star, three: Star, four: Star): Hint =
        Hint.create(description, one, two, three, four).also {
            universe.hints.add(it)
        }


    fun newSatellite(name: String, displayName: String, noradSatelliteNumber: Int, elements: String): Satellite =
        universe.satellites.findSatelliteByName(name) ?:
            Satellite.createNewSatellite(name, displayName, noradSatelliteNumber, elements).also {
                universe.satellites.add(it)
            }

    fun newSpecialElement(name: String, ascension: Hour, declination: Degree): SpecialElement =
        SpecialElement(name, ascension.angleInDegree, declination.angleInDegree).also {
            universe.specials.add(it)
        }

    fun setFriendlyNameTo(hd: Int, friendlyName: String): Star =
        getStar(hd).also {
            universe.stars.setFriendlyName(it, friendlyName)
        }

    fun asVIP(star: Star) {
        if (!universe.vip.contains(star)) {
            universe.vip.add(star)
        }
    }
}

